### ISSUE TITLE
Add --optimistic-linking to GHC RTS Flags exposed in base.

### ISSUE THREAD

**Issue by doyougnu** [haskell/core-libraries-committee#288]
### What 

This is part of the Haskell.nix upstream project. One patch in that project patched GHC's runtime linker to be much less pessimistic, allowing it to continue linking to the best of its ability in the presence of unknown symbols. This feature has now been implemented in GHC (See implementation) but is hidden behind a new RTSMisc flag called `--optimistic-linking`. Because this changes the RTS flags, and these flags are exposed in base, this CLC proposal was requested.

### Why

There are two major reasons for this feature:

1. The original reason: for some targets, like iOS and Android linking is more complex and this feature allows linking even if a symbol is present but is never used. See Moritz's comments in the GHC MR linked in Implementation.

2. The maintenance reason: close the gap between Haskell.nix and GHC upstream. 

### Implementation

MR is here: https://gitlab.haskell.org/ghc/ghc/-/merge_requests/13012

### Impact

Should be similar to #243 #254, and #263 and the impact listed in that proposal. Note that the default behavior of the RTS linker is unchanged.

**Comment by hasufell** [comment]
Quoting @dcoutts from https://github.com/haskell/core-libraries-committee/issues/263

> Longer term it would seem wise for GHC.RTS.Flags to not be exported from base at all since it is inherently GHC-specific and tightly coupled to internal details of the implementation

Adding a record is a breaking change. These repeated MRs cause churn for base.

I'm inclined to vote -1 and wait for someone to remove this module from base. Because it seems we're not gonna make any progress on it otherwise.

**Comment by doyougnu** [comment]
> How feasible is the removal of `GHC.RTS.Flags` with the current state of `ghc-internal`? This should really be a part of the compiler library rather than the standard library. I would also be much more receptive to a breaking change of removing this module and pointing its sparse users elsewhere.

Opened #289 for this discussion. I am open to working on that project.

**Comment by doyougnu** [comment]
Now that #289 is open. I would like to merge this but hide the flag during the re-export in base. This way this proposal can still move forward without waiting on the result of #289.

**Comment by maralorn** [comment]
> Now that #289 is open. I would like to merge this but hide the flag during the re-export in base. This way this proposal can still move forward without waiting on the result of #289.

Isn’t that a change which simply does not require a CLC vote? Or would this still be an API change for base but one less likely to break anything?

**Comment by adamgundry** [comment]
I think merely hiding the field selector is problematic, because the data constructor is exported?

It would be possible to change the type in `ghc-internal` without making an API change to `base`, but that would require a bit of shimming (and if anyone is very pedantic, the shim might constitute a change to the performance of the API and hence require CLC approval...).

**Comment by Bodigrim** [comment]
> I think merely hiding the field selector is problematic, because the data constructor is exported?

Correct. It remains a breaking change, so I'm not sure there is a tangible benefit from hiding the record selector.

----

I'm personally pretty much at ease with updating `GHC.RTS.Flags` frequently. After all it says "The API of this module is unstable" for a reason. As long as we adhere to PVP (which we do), I'm fine with it. Indeed it's a bit tedious for both parties to run CLC proposals each time, so I'd love to see #289 resolved instead, but in principle I'd vote +1 on this proposal too.

**Comment by doyougnu** [comment]
I'm closing this as superseded by #289. The MR linked in the proposal has been changed to no longer export the flag to base.
